json/pickler/src/main/scala/sttp/tapir/json/Pickler.scala

```diff
@@ -0,0 +1,359 @@
+import sttp.tapir.DecodeResult.{Error, Value}
+import sttp.tapir.SchemaType.SProduct
+import sttp.tapir.generic.Configuration
+import sttp.tapir.{Codec, Schema, SchemaAnnotations, Validator}
+
+import scala.collection.Factory
+import scala.compiletime.*
+import scala.deriving.Mirror
+import scala.quoted.*
+import scala.reflect.ClassTag
+import scala.util.{Failure, NotGiven, Success, Try}
+
+import macros.*
+
+object Pickler:
+
+  inline def derived[T: ClassTag](using Configuration, Mirror.Of[T]): Pickler[T] =
+    given subtypeDiscriminator: SubtypeDiscriminator[T] = DefaultSubtypeDiscriminator()
+    summonFrom {
+      case schema: Schema[T] => fromExistingSchemaAndRw[T](schema)
+      case _                 => buildNewPickler[T]()
+    }
+
+  inline def oneOfUsingField[T: ClassTag, V](extractor: T => V, asString: V => String)(
+      mapping: (V, Pickler[_ <: T])*
+  )(using m: Mirror.Of[T], c: Configuration, p: Pickler[V]): Pickler[T] =
+
+    val paramExtractor = extractor
+    val paramAsString = asString
+    val paramMapping = mapping
+    type ParamV = V
+    given subtypeDiscriminator: SubtypeDiscriminator[T] = new CustomSubtypeDiscriminator[T] {
+      type V = ParamV
+      override def extractor = paramExtractor
+      override def asString = paramAsString
+      override lazy val mapping = paramMapping
+    }
+    summonFrom {
+      case schema: Schema[T] => fromExistingSchemaAndRw[T](schema)
+      case _ =>
+        inline m match {
+          case p: Mirror.ProductOf[T] =>
+            error(
+              s"Unexpected product type (case class) ${implicitly[ClassTag[T]].runtimeClass.getSimpleName()}, this method should only be used with sum types (like sealed hierarchy)"
+            )
+          case _: Mirror.SumOf[T] =>
+            inline if (isScalaEnum[T])
+              error("oneOfUsingField cannot be used with enums. Try Pickler.derivedEnumeration instead.")
+            else {
+              given schemaV: Schema[V] = p.schema
+              val schema: Schema[T] = Schema.oneOfUsingField[T, V](extractor, asString)(
+                mapping.toList.map { case (v, p) =>
+                  (v, p.schema)
+                }: _*
+              )
+              lazy val childPicklers: Tuple.Map[m.MirroredElemTypes, Pickler] = summonChildPicklerInstances[T, m.MirroredElemTypes]
+              picklerSum(schema, childPicklers)
+            }
+        }
+    }
+
+  inline def derivedEnumeration[T: ClassTag](using Mirror.Of[T]): CreateDerivedEnumerationPickler[T] =
+    inline erasedValue[T] match
+      case _: Null =>
+        error("Unexpected non-enum Null passed to derivedEnumeration")
+      case _: Nothing =>
+        error("Unexpected non-enum Nothing passed to derivedEnumeration")
+      case _: reflect.Enum =>
+        new CreateDerivedEnumerationPickler(Validator.derivedEnumeration[T], SchemaAnnotations.derived[T])
+      case other =>
+        error(s"Unexpected non-enum value $other passed to derivedEnumeration")
```

Comment:
shouldn't this be `non-enum type`, not value? also, why are Null/Nothing special-cased?

Reply:
It turns out `Null` and `Nothing` will be matched against `case _: reflect.Enum`. In this case we have an additional constraint `T: ClassTag`, which prevents using these types, it's also super unlikely that anyone would fall into these cases anyway, so we can remove these cases.
As for the error message, the `$other` variable is actually a value, but it can't be put into a string  in this way anyway, so I'll just change the message to `Unexpected non-enum type passed to derivedEnumeration`. It's obvious which type is problematic here:
```
[error] -- Error: /home/kc/code/oss/tapir/json/pickler/src/test/scala/sttp/tapir/json/PicklerTest.scala:508:4
[error] 508 |    Pickler.derivedEnumeration[Book]
[error]     |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |    Unexpected non-enum type passed to derivedEnumeration
```